File: backend/src/core/SecurityBot.java
package core;

import java.util.*;
import model.LogEntry;
import alert.AlertSystem;
import service.LogAnalyzer;

public class SecurityBot {

    private Map<String, Integer> dangerHistory = new HashMap<>();

    /**
     * Số lượng LOGIN_FAIL và REQUEST đã được phân tích ở chu kỳ trước.
     * dangerHistory chỉ tăng khi fail/request count TĂNG LÊN (có log mới).
     * Nếu count giữ nguyên (cùng log cũ trong DB), không tăng history.
     */
    private Map<String, Integer> lastFailCount = new HashMap<>();
    private Map<String, Integer> lastRequestCount = new HashMap<>();

    /**
     * Firewall instance — blocks IPs that exceed the risk threshold.
     */
    private Firewall firewall = new Firewall();

    /**
     * Analyzes a list of log entries and assigns risk scores, statuses,
     * and attack types to each entry based on IP behavior.
     *
     * @param logs        List of network log events (LogEntry).
     * @param alertSystem AlertSystem used to emit alert messages.
     */
    public synchronized void analyze(List<LogEntry> logs, AlertSystem alertSystem) {

        LogAnalyzer analyzer = new LogAnalyzer();

        // Step 1: Count actions per IP
        Map<String, Integer> failMap = analyzer.countAction(logs, "LOGIN_FAIL");
        Map<String, Integer> requestMap = analyzer.countAction(logs, "REQUEST");
        Map<String, Integer> successMap = analyzer.countAction(logs, "LOGIN_SUCCESS");

        // Step 2: Collect all unique IPs seen across all action maps
        Set<String> allIPs = new HashSet<>();
        allIPs.addAll(failMap.keySet());
        allIPs.addAll(requestMap.keySet());
        allIPs.addAll(successMap.keySet());
        // Giữ lại state cho các IP cũ (tránh reset về 0/-1)
        Set<String> existingIPs = new HashSet<>(lastFailCount.keySet());

        for (String oldIp : existingIPs) {
            if (!allIPs.contains(oldIp)) {
                // giữ nguyên giá trị cũ, KHÔNG xóa
                lastFailCount.put(oldIp, lastFailCount.get(oldIp));
                lastRequestCount.put(oldIp, lastRequestCount.getOrDefault(oldIp, 0));
            }
        }
        // Step 3: Evaluate each IP
        for (String ip : allIPs) {

            if (firewall.isBlocked(ip)) {
                for (LogEntry log : logs) {
                    if (log.getIp().equals(ip)) {
                        log.setScore(90);
                        log.setStatus("BLOCKED");
                        if (log.getAttackType() == null || log.getAttackType().equals("NORMAL")) {
                            log.setAttackType("BLOCKED");
                        }
                    }
                }
                continue;
            }

            int fail = failMap.getOrDefault(ip, 0);
            int request = requestMap.getOrDefault(ip, 0);
            int success = successMap.getOrDefault(ip, 0);
            int history = dangerHistory.getOrDefault(ip, 0);

            // So sánh với snapshot chu kỳ trước:
            // - Default -1: lần đầu thấy IP, LUÔN hasNewActivity=true → detect đúng ngay
            // cycle 1.
            // (fail=2 > -1 → true → alert 1 lần đúng lúc)
            // - Cycle 2+: prevFail=2, fail=2 → (2>2)=false → NO alert ← đây là điểm mấu
            // chốt.
            // containsKey() bị bỏ vì: có blind spot 2s đầu và bị re-trigger khi log
            // ra/vào top-100 window (prevFail bị reset về 0, rồi fail=2 > 0 → sai).
            int prevFail = lastFailCount.getOrDefault(ip, 0);
            int prevRequest = lastRequestCount.getOrDefault(ip, 0);
            boolean hasNewActivity = (fail > prevFail) || (request > prevRequest);

            // Luôn cập nhật snapshot cho chu kỳ tiếp theo
            lastFailCount.put(ip, fail);
            lastRequestCount.put(ip, request);

            // Risk score formula:
            // LOGIN_FAIL x10 — strong signal of brute force
            // REQUEST x5 — flood traffic
            // SUCCESS x1 — reduces suspicion
            // history x5 — prior offense multiplier (chỉ tăng khi có activity mới)
            int riskScore = (fail * 10)
                    + (request * 5)
                    - (success * 1)
                    + (history * 5);
            riskScore = Math.max(riskScore, 0);

            // Step 4: Determine attack type
            String attackType = "NORMAL";
            if (fail >= 1) {
                attackType = "BRUTE_FORCE";
            } else if (request >= 1) {
                attackType = "REQUEST_FLOOD";
            }

            // Step 5: Chỉ leo thang (history, alert, block) khi có LOG MỚI.
            // Không có log mới → cùng bộ log cũ trong DB → không tạo thêm alert,
            // không tự block, không tăng history. Status giữ nguyên nhờ no-downgrade rule.
            String status = "PASS";

            if (riskScore >= 15 && hasNewActivity) {
                int level = Math.min(history + 1, 10);
                dangerHistory.put(ip, level);

                if (riskScore >= 45) {
                    firewall.blockIP(ip);
                    status = "BLOCKED";

                } else if (riskScore >= 30) {
                    status = "MONITORING";

                } else {
                    status = "SUSPICIOUS";
                }

                alertSystem.addAlert(ip, attackType, riskScore, status);
            }

            // Step 6: Cập nhật score và status lên các log entries của IP.
            // status="PASS" khi không có activity mới → no-downgrade rule giữ nguyên
            // SUSPICIOUS/MONITORING hiện tại, không downgrade.
            updateLogsForIP(logs, ip, riskScore, status, attackType);
        }
    }

    /*
     * 
     * Updates score, status, and attackType on all log entries belonging to a given
     * IP.
     * 
     * 
     * 
     * @param logs Full list of log entries
     * 
     * @param ip IP address to match
     * 
     * @param riskScore Computed risk score
     * 
     * @param status Status label (PASS / SUSPICIOUS / MONITORING / BLOCKED)
     * 
     * @param attackType Attack type label (NORMAL / BRUTE_FORCE / REQUEST_FLOOD)
     */
    private void updateLogsForIP(List<LogEntry> logs, String ip,
            int riskScore, String status, String attackType) {
        for (LogEntry log : logs) {
            if (log.getIp().equals(ip)) {

                int currentScore = Math.max(log.getScore(), riskScore);
                String currentStatus = log.getStatus();

                // Mỗi log giữ nguyên attack_type gốc được set lúc INSERT:
                // - LOGIN_FAIL log → BRUTE_FORCE (set bởi /api/attack/bruteforce)
                // - REQUEST log → REQUEST_FLOOD (set bởi /api/attack/flood)
                // - Log thường → NORMAL
                // attackType (tham số) chỉ dùng cho alertSystem.addAlert() ở trên,
                // không áp đặt lên từng bản ghi DB nữa.

                // Luật chống hạ cấp status:
                // Không cho phép ghi đè "PASS" lên log đang ở mức SUSPICIOUS/MONITORING
                if (status.equals("PASS")
                        && (currentStatus.equals("SUSPICIOUS") || currentStatus.equals("MONITORING"))) {
                    // Giữ nguyên status lịch sử, không downgrade
                } else {
                    log.setStatus(status);
                }

                log.setScore(currentScore);
            }
        }
    }

    /**
     * Xóa lịch sử nguy hiểm (dangerHistory) của một IP cụ thể.
     *
     * Được gọi khi admin gỡ block IP từ /api/unblock.
     * Nếu không xóa dangerHistory, analyze() sẽ đọc score lịch sử cao
     * và re-block IP sau ~2 giây ngay cả khi đã gỡ block.
     *
     * @param ip IP address cần xóa lịch sử
     */
    public synchronized void clearHistory(String ip) {
        dangerHistory.remove(ip);
        lastFailCount.remove(ip); // Reset snapshot để IP bắt đầu lại hoàn toàn
        lastRequestCount.remove(ip);
        System.out.println("[SecurityBot] Cleared danger history for IP: " + ip);
    }

    /**
     * Returns the Firewall instance (used by ApiServer to expose blocked IPs).
     */
    public Firewall getFirewall() {
        return firewall;
    }
}

File: backend/core/security_bot.py
import threading
from typing import List, Dict, Set
from models.log_entry import LogEntry
from core.firewall import Firewall
from alert.alert_system import AlertSystem
from services.log_analyzer import LogAnalyzer
from core.ai_detector import AIDetector

class SecurityBot:
    """
    SecurityBot — Evaluates log entries, calculates risk scores,
    manages danger history, triggers alerts, and interacts with the Firewall.
    Tích hợp AIDetector để nhận diện thông minh các cuộc tấn công Web và luồng mạng.
    """

    def __init__(self):
        self.danger_history: Dict[str, int] = {}
        self.last_fail_count: Dict[str, int] = {}
        self.last_request_count: Dict[str, int] = {}
        self.last_alert_score: Dict[str, int] = {}
        self.firewall = Firewall()
        self.ai_detector = AIDetector()
        self._lock = threading.Lock()

    def analyze(self, logs: List[LogEntry], alert_system: AlertSystem):
        with self._lock:
            analyzer = LogAnalyzer()

            fail_map = analyzer.count_action(logs, "LOGIN_FAIL")
            request_map = analyzer.count_action(logs, "REQUEST")
            success_map = analyzer.count_action(logs, "LOGIN_SUCCESS")

            all_ips: Set[str] = set()
            all_ips.update(fail_map.keys())
            all_ips.update(request_map.keys())
            all_ips.update(success_map.keys())

            existing_ips = set(self.last_fail_count.keys())
            for old_ip in existing_ips:
                if old_ip not in all_ips:
                    self.last_fail_count[old_ip] = self.last_fail_count.get(old_ip, 0)
                    self.last_request_count[old_ip] = self.last_request_count.get(old_ip, 0)

            for ip in all_ips:
                if self.firewall.is_blocked(ip):
                    for log in logs:
                        if log.get_ip() == ip:
                            log.set_score(90)
                            log.set_status("BLOCKED")
                            if not log.get_attack_type() or log.get_attack_type() == "NORMAL":
                                log.set_attack_type("BLOCKED")
                    continue

                fail = fail_map.get(ip, 0)
                request = request_map.get(ip, 0)
                success = success_map.get(ip, 0)
                history = self.danger_history.get(ip, 0)

                prev_fail = self.last_fail_count.get(ip, 0)
                prev_request = self.last_request_count.get(ip, 0)
                has_new_activity = (fail > prev_fail) or (request > prev_request)

                self.last_fail_count[ip] = fail
                self.last_request_count[ip] = request

                # ── Phân tích thông minh bằng AI Detector ──
                ai_attack_type = None
                ai_max_risk = 0
                for log in logs:
                    if log.get_ip() == ip:
                        desc = log.get_description()
                        if desc and desc.strip():
                            ai_res = self.ai_detector.predict_payload(desc)
                            if ai_res.get("is_attack"):
                                ai_attack_type = f"WEB_{ai_res['attack_type'].upper()}"
                                ai_max_risk = max(ai_max_risk, ai_res.get("risk_score", 0))
                                log.set_attack_type(ai_attack_type)
                                log.set_score(max(log.get_score(), ai_res.get("risk_score", 0)))

                risk_score = (fail * 10) + (request * 5) - (success * 1) + (history * 5)
                risk_score = max(risk_score, ai_max_risk, 0)

                attack_type = "NORMAL"
                if ai_attack_type:
                    attack_type = ai_attack_type
                elif fail >= 5:
                    attack_type = "BRUTE_FORCE"
                elif request >= 20:
                    attack_type = "REQUEST_FLOOD"

                status = "PASS"
                prev_alert_score = self.last_alert_score.get(ip, 0)

                if risk_score >= 15 and (has_new_activity or ai_attack_type):
                    level = min(history + 1, 10)
                    self.danger_history[ip] = level

                    if risk_score >= 45:
                        self.firewall.block_ip(ip)
                        status = "BLOCKED"
                    elif risk_score >= 30:
                        status = "MONITORING"
                    else:
                        status = "SUSPICIOUS"

                    if risk_score > prev_alert_score:
                        alert_system.add_alert(ip, attack_type, risk_score, status)
                        self.last_alert_score[ip] = risk_score

                self._update_logs_for_ip(logs, ip, risk_score, status, attack_type)

    def _update_logs_for_ip(self, logs: List[LogEntry], ip: str,
                            risk_score: int, status: str, attack_type: str):
        for log in logs:
            if log.get_ip() == ip:
                current_score = max(log.get_score(), risk_score)
                current_status = log.get_status()

                if status == "PASS" and current_status in ("SUSPICIOUS", "MONITORING"):
                    pass
                else:
                    log.set_status(status)

                log.set_score(current_score)

    def clear_history(self, ip: str):
        with self._lock:
            self.danger_history.pop(ip, None)
            self.last_fail_count.pop(ip, None)
            self.last_request_count.pop(ip, None)
            self.last_alert_score.pop(ip, None)
            print(f"[SecurityBot] Cleared danger history for IP: {ip}")

    def get_firewall(self) -> Firewall:
        return self.firewall

    def get_ai_detector(self) -> AIDetector:
        return self.ai_detector
